terra_notebook_utils/utils.py

```diff
@@ -0,0 +1,34 @@
+from concurrent.futures import ThreadPoolExecutor, Future, as_completed
+from typing import Any, Callable, Set
+
+
+class _AsyncContextManager:
+    """
+    Context manager for asynchronous execution. Wait on exit for all jobs to complete.
+    """
+    def submit(self, func: Callable[..., Any], *args, **kwargs):
+        f = self._executor.submit(func, *args, **kwargs)
+        self._futures.add(f)
+        self.prune_futures()
+
+    def prune_futures(self):
+        for f in self._futures.copy():
+            if f.done():
+                f.result()
+                self._futures.remove(f)
+
+    def _prepare_for_exit(self):
+        raise NotImplementedError()
+
+    def __enter__(self):
+        self._executor = ThreadPoolExecutor(max_workers=10)
```

Comment:
Should `max_workers` be proportional to max_cpu?

Reply:
Nope: this is an IO bound task, so it shouldn’t be set to the number of CPUs. I’m actually not sure what it should be set to. In the meantime, 10 seems like a decent guess according to empirical evidence.